File: 3327.check-if-dfs-strings-are-palindromes.cpp
#
# @lc app=leetcode id=3327 lang=cpp
#
# [3327] Check if DFS Strings Are Palindromes
#

# @lc code=start
class Solution {
public:
    vector<bool> findAnswer(vector<int>& parent, string s) {
        int n = s.size();
        vector<vector<int>> adj(n);
        for (int i = 1; i < n; ++i) {
            adj[parent[i]].push_back(i);
        }
        for (auto& ch : adj) {
            sort(ch.begin(), ch.end());
        }
        const long long MOD1 = 1000000007LL;
        const long long MOD2 = 1000000009LL;
        const long long B1 = 31LL;
        const long long B2 = 37LL;
        vector<long long> p1(n + 1, 1LL);
        vector<long long> p2(n + 1, 1LL);
        for (int i = 1; i <= n; ++i) {
            p1[i] = p1[i - 1] * B1 % MOD1;
            p2[i] = p2[i - 1] * B2 % MOD2;
        }
        vector<long long> hf1(n), hf2(n), hr1(n), hr2(n);
        vector<int> len(n);
        std::function<void(int)> dfs = [&](int u) {
            len[u] = 1;
            long long f1 = 0, f2 = 0;
            long long r1 = (s[u] - 'a' + 1) % MOD1;
            long long r2 = (s[u] - 'a' + 1) % MOD2;
            for (int v : adj[u]) {
                dfs(v);
                len[u] += len[v];
                f1 = (f1 * p1[len[v]] % MOD1 + hf1[v]) % MOD1;
                f2 = (f2 * p2[len[v]] % MOD2 + hf2[v]) % MOD2;
            }
            long long ch = s[u] - 'a' + 1;
            f1 = (f1 * p1[1] % MOD1 + ch) % MOD1;
            f2 = (f2 * p2[1] % MOD2 + ch) % MOD2;
            for (int j = adj[u].size() - 1; j >= 0; --j) {
                int v = adj[u][j];
                r1 = (r1 * p1[len[v]] % MOD1 + hr1[v]) % MOD1;
                r2 = (r2 * p2[len[v]] % MOD2 + hr2[v]) % MOD2;
            }
            hf1[u] = f1;
            hf2[u] = f2;
            hr1[u] = r1;
            hr2[u] = r2;
        };
        dfs(0);
        vector<bool> ans(n);
        for (int i = 0; i < n; ++i) {
            ans[i] = (hf1[i] == hr1[i] && hf2[i] == hr2[i]);
        }
        return ans;
    }
};
# @lc code=end

File: 3327.check-if-dfs-strings-are-palindromes.py
#
# @lc app=leetcode id=3327 lang=python3
#
# [3327] Check if DFS Strings Are Palindromes
#

# @lc code=start
from typing import List
class Solution:
    def findAnswer(self, parent: List[int], s: str) -> List[bool]:
        n = len(parent)
        # Build adjacency list
        tree = [[] for _ in range(n)]
        for i in range(1, n):
            tree[parent[i]].append(i)
        for node in range(n):
            tree[node].sort()  # Ensure increasing order traversal

        # Hash parameters
        MOD1, MOD2 = 10**9 + 7, 10**9 + 9
        BASE1, BASE2 = 911, 3571

        # Precompute powers for both hashes
        pow1 = [1] * (n + 2)
        pow2 = [1] * (n + 2)
        for i in range(1, n + 2):
            pow1[i] = (pow1[i-1] * BASE1) % MOD1
            pow2[i] = (pow2[i-1] * BASE2) % MOD2

        # Store hashes and lengths for each node
        fw_hash1 = [0] * n
        fw_hash2 = [0] * n
        rv_hash1 = [0] * n
        rv_hash2 = [0] * n
        length = [0] * n

        def dfs(x):
            h1, h2 = 0, 0
            rh1, rh2 = 0, 0
            l = 0
            for y in tree[x]:
                dfs(y)
                # Append child's hashes/lengths to current
                h1 = (h1 * pow1[length[y]] + fw_hash1[y]) % MOD1
                h2 = (h2 * pow2[length[y]] + fw_hash2[y]) % MOD2
                rh1 = (rv_hash1[y] * pow1[l] + rh1) % MOD1
                rh2 = (rv_hash2[y] * pow2[l] + rh2) % MOD2
                l += length[y]
            # Add current node's character
            code = ord(s[x]) - ord('a') + 1
            h1 = (h1 * BASE1 + code) % MOD1
            h2 = (h2 * BASE2 + code) % MOD2
            rh1 = (rh1 + code * pow1[l]) % MOD1
            rh2 = (rh2 + code * pow2[l]) % MOD2
            l += 1
            fw_hash1[x], fw_hash2[x] = h1, h2
            rv_hash1[x], rv_hash2[x] = rh1, rh2
            length[x] = l

        for i in range(n):
            length[i] = 0
        dfs(0)

        answer = [False] * n
        def dfs_check(x):
            # For the subtree rooted at x, check palindrome
            answer[x] = (fw_hash1[x] == rv_hash1[x]) and (fw_hash2[x] == rv_hash2[x])
            for y in tree[x]:
                dfs_check(y)
        dfs_check(0)
        return answer
# @lc code=end